Source/WebCore/platform/graphics/skia/SkiaThreadedPaintingPool.cpp

```diff
@@ -60,30 +75,92 @@ std::unique_ptr<DisplayList::DisplayList> SkiaThreadedPaintingPool::recordDispla
+    context.drawDisplayListItems(displayList.items(), displayList.resourceHeap(), *s_controlFactory, FloatPoint::zero());
+
+    canvas->restore();
+    return true;
+}
+
+bool SkiaThreadedPaintingPool::paintGraphicsLayerIntoBuffer(Ref<Nicosia::Buffer>& buffer, const TiledBackingStore& tiledBackingStore, const CoordinatedGraphicsLayer& layer, const IntRect& dirtyRect) const
+{
+    auto* canvas = buffer->canvas();
+    if (!canvas)
+        return false;
+
+    canvas->save();
+    canvas->clear(SkColors::kTransparent);
+
+    GraphicsContextSkia context(*canvas, buffer->isBackedByOpenGL() ? RenderingMode::Accelerated : RenderingMode::Unaccelerated, RenderingPurpose::LayerBacking);
+    layer.paintIntoGraphicsContext(context, tiledBackingStore, dirtyRect);
+
+    canvas->restore();
+    return true;
+}
+
+bool SkiaThreadedPaintingPool::canPerformAcceleratedRendering() const
+{
+    return ProcessCapabilities::canUseAcceleratedBuffers() && PlatformDisplay::sharedDisplay().skiaGLContext()->makeContextCurrent();
+}
+
+bool SkiaThreadedPaintingPool::canPerformThreadedAcceleratedRendering() const
+{
+    return !!m_gpuWorkerPool && canPerformAcceleratedRendering();
+}
+
+bool SkiaThreadedPaintingPool::canPerformThreadedUnacceleratedRendering() const
+{
+    return !!m_cpuWorkerPool;
+}
+
+Ref<Nicosia::Buffer> SkiaThreadedPaintingPool::createBuffer(RenderingMode renderingMode, const CoordinatedGraphicsLayer& layer, const IntSize& size) const
+{
+    if (renderingMode == RenderingMode::Accelerated) {
+        OptionSet<BitmapTexture::Flags> textureFlags;
+        if (!layer.contentsOpaque())
+            textureFlags.add(BitmapTexture::Flags::SupportsAlpha);
+
+        ASSERT(m_texturePool);
+        return Nicosia::AcceleratedBuffer::create(m_texturePool->acquireTexture(size, textureFlags));
+    }
+
+    return Nicosia::UnacceleratedBuffer::create(size, layer.contentsOpaque() ? Nicosia::Buffer::NoFlags : Nicosia::Buffer::SupportsAlpha);
+}
+
+Ref<Nicosia::Buffer> SkiaThreadedPaintingPool::postPaintingTask(RenderingMode renderingMode, const TiledBackingStore& tiledBackingStore, const CoordinatedGraphicsLayer& layer, const IntRect& dirtyRect)
 {
     WTFBeginSignpost(this, RecordTile);
     auto displayList = recordDisplayList(tiledBackingStore, layer, dirtyRect);
     WTFEndSignpost(this, RecordTile);
 
+    auto buffer = createBuffer(renderingMode, layer, dirtyRect.size());
     buffer->beginPainting();
 
-    m_workerPool->postTask([buffer = Ref { buffer }, displayList = WTFMove(displayList), dirtyRect]() mutable {
-        if (auto* canvas = buffer->canvas()) {
-            canvas->save();
-            canvas->clear(SkColors::kTransparent);
-
-            static thread_local RefPtr<ControlFactory> s_controlFactory;
-            if (!s_controlFactory)
-                s_controlFactory = ControlFactory::create();
+    auto* workerPool = renderingMode == RenderingMode::Accelerated ? m_gpuWorkerPool.get() : m_cpuWorkerPool.get();
```

Comment:
Better use a RefPtr here instead of  a raw pointer

Reply:
Why unnecessary ref'ing? We're only using it the line after.

Reply:
Just to void raw pointers, you can also move the assert before and then use a reference.

Reply:
Fair enough, let's do that.